internal/resolver/dns/dns_resolver_test.go

```diff
@@ -1290,3 +1290,45 @@ func (s) TestMinResolutionInterval(t *testing.T) {
 		r.ResolveNow(resolver.ResolveNowOptions{})
 	}
 }
+
+// TestMinResolutionInterval_NoExtraDelay verifies that there is no extra delay
+// between two resolution requests apart from [MinResolutionInterval]
+// i.e. if a request is made after [MinResolutionInterval], it should resolve
+// immediately. Test sets [MinResolutionInterval] to 1 second and calls
+// ResolveNow() 4 times with a wait of 1 second between each request.
+// A correct implementation should resolve all requests without timing out in
+// test duration of 5 seconds.
+func (s) TestMinResolutionInterval_NoExtraDelay(t *testing.T) {
+	const target = "foo.bar.com"
+
+	overrideResolutionInterval(t, 1*time.Second)
+	tr := &testNetResolver{
+		hostLookupTable: map[string][]string{
+			"foo.bar.com": {"1.2.3.4", "5.6.7.8"},
+		},
+		txtLookupTable: map[string][]string{
+			"_grpc_config.foo.bar.com": txtRecordServiceConfig(txtRecordGood),
+		},
+	}
+	overrideNetResolver(t, tr)
+
+	r, stateCh, _ := buildResolverWithTestClientConn(t, target)
+
+	wantAddrs := []resolver.Address{{Addr: "1.2.3.4" + colonDefaultPort}, {Addr: "5.6.7.8" + colonDefaultPort}}
+	wantSC := scJSON
+
+	// set context timeout to test duration of 5 seconds
+	// to make sure all 4 resolutions happen successfully
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	// We chose to make 4 requests to strike a balance between coverage and
+	// test duration. This number is sufficiently large to validate the
+	// behavior across multiple resolution attempts, while also reducing
+	// the likelihood of flakiness due to timing issues.
+	for i := 0; i < 4; i++ {
+		verifyUpdateFromResolver(ctx, t, stateCh, wantAddrs, nil, wantSC)
+		time.Sleep(1 * time.Second) // respect resolution rate of 1 second for re-resolve
```

Comment:
We try to avoid sleeping in our tests as much as possible. Sleeping causes the tests to unnecessarily flaky, and run for longer than actually required.

Instead, if you override `internal.TimeAfterFunc`, you can control exactly when this function pushes something on the channel read by the code. This way, you can avoid sleeping as well.

There is currently an `overrideTimeAfterFunc` helper function, but this sets the `after func` to be invoked after a set duration. Instead you can override `internal.TimeAfterFunc` to return a channel that is controlled by the test, and have it push to it whenever it wants the code under test to be unblocked.

Reply:
I don't think we should overriding `internal.TimeAfterFunc` because we want to test the actual implementation work correctly with `MinResolutionInterval`. Here, we just want to wait outside for 1 second before making another request so I just replaced `time.Sleep` with a channel using `time.After`

Reply:
No, we don't want this test to run for 4+ seconds.

Instead I suggest the following:
- Define a variable for `internal.TimeNowFunc`, just like how it is done for `internal.TimeAfterFunc` 
  - set this variable to `time.Now` in the `init` function
- Override this variable to return a time value that you control in your test
- Override `internal.TimeAfterFunc` in your test and verify that it is called with the expected duration

Let me know if this doesn't work.

Reply:
Thanks for suggestion. Modified the test to override `time.Until` to calculate remaining wait to allow resolution after 1s i.e. equivalent of making re-resolution after 1s and verify the expected wait duration is very small.